### PR TITLE
Implement AdminEmailLogPageActionTest #6855

### LINKED ISSUE #6855: Implement AdminEmailLogPageActionTest
Currently, `AdminEmailLogPageActionTest.java` is just missing. Since from #6655, we can manually inject log messages. Therefore, the test case is doable.

### PR DESCRIPTION
**Pull request by xpdavid**
Fixes #6855

**Outline of Solution**

> 1. Fix several bugs in AdminEmailLogPageAction
> 1. Implement AdminEmailLogPageActionTest

This one would be similar to #6655

### CODE REVIEW

**Comment by junhaoyap** [comment]
@xpdavid travis build is failing

**Comment by CanIHasReview** [comment]
# v1

@xpdavid submitted v1 for review.

* [v1 1/6] [AdminEmailLogPageData: fix bugs: isShowTestData is ignored when query is invalid ](https://github.com/TEAMMATES/teammates/pull/6876/commits/ff563f48b02a834e70f31ca265af59359ace17ba)
* [v1 2/6] [AdminEmailLog: Remove dead code \(cannot find usage of these codes\) ](https://github.com/TEAMMATES/teammates/pull/6876/commits/7fe5dc11d74baadb31637d06125d0eaa7a9fc7e3)
* [v1 3/6] [AdminEmailLog: Fix bugs for continue search with query \`after\` and \`before\` ](https://github.com/TEAMMATES/teammates/pull/6876/commits/a036d985e3fe24c22b47465c5d20fa0f568f8161)
* [v1 4/6] [AdminEmailLog: Ignore whitespaces in query ](https://github.com/TEAMMATES/teammates/pull/6876/commits/76565edea24234328af0cb44307d5f7869627628)
* [v1 5/6] [AdminActivityLogTest: Rename group name to avoid confusion ](https://github.com/TEAMMATES/teammates/pull/6876/commits/936c9e31cef574e1be79081584755f2772ede2d0)
* [v1 6/6] [Implement AdminEmailLogPageActionTest ](https://github.com/TEAMMATES/teammates/pull/6876/commits/73f8060934462c5d98cb324a510c303d165b29f2)

([:books: Archive](https://github.com/TEAMMATES/teammates/compare/4cd942a04b34aea9ac58702b49243c18befb9374...73f8060934462c5d98cb324a510c303d165b29f2))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/TEAMMATES/teammates.git refs/pr/6876/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/main/java/teammates/ui/pagedata/AdminEmailLogPageData.java [hunk 1]
```diff
@@ -168,21 +203,17 @@ private QueryParameters parseQuery(String query) throws ParseException, InvalidP
...
             }
-            return true;
+            return isShowTestData;
         }
 
         // filter based on what is in the query
-        if (q.isToDateInQuery && logEntry.getTime() > q.toDateValue) {
-            return false;
-        }
-        if (q.isFromDateInQuery && logEntry.getTime() < q.fromDateValue) {
```
Comment (whipermr5) [inline src/main/java/teammates/ui/pagedata/AdminEmailLogPageData.java]: Are these lines not needed to hide the log entry if it does not fall within the user-specified `before:` `after:` range?

File: src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+        AdminEmailLogPageData pageData = (AdminEmailLogPageData) result.data;
+        List<EmailLogEntry> actualLogs = getLogsFromLogTemplateRows(pageData.getLogs());
+        verifyLogs(expectedLogs, actualLogs);
+    }
+
+    /**
+     * Verifies actualLogs contains expectedLogs.
+     *
+     * <p>expectedLogs is a 2D array, the outer indices correspond to {@link #LOG_MESSAGE_INDEX_TODAY}
+     * {@link #LOG_MESSAGE_YESTDAY_INDEX} and {@link #LOG_MESSAGE_INDEX_TWO_DAYS_AGO}, the inner indices for
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java]: `LOG_MESSAGE_YESTDAY_INDEX` -> `LOG_MESSAGE_INDEX_YESTERDAY`

File: src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+        query = "info:";
+        verifyActionResult(expected, "filterQuery", query);
+        query = "receiver unknown_connector info:content";
+        verifyActionResult(expected, "filterQuery", query);
+        query = "unknown:subject1 | info:keyword1";
+        verifyActionResult(expected, "filterQuery", query);
+
+        // invalid filterQuery with showing testing data
+        expected = new int[][]{{0, 1, 2, 3, 4, 5, 6}};
+        query = "information:unkown";
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java]: Is `unkown` a typo?

File: src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+        insertLogMessagesAtTime(logMessages.get(LOG_MESSAGE_INDEX_YESTERDAY), yesterday.getTime());
+        Date today = TimeHelper.getDateOffsetToCurrentTime(0);
+        insertLogMessagesAtTime(logMessages.get(LOG_MESSAGE_INDEX_TODAY), today.getTime());
+    }
+
+    @Test(groups = "typicalEmailLogs")
+    public void filterQuery_invalidQuery_defaultSearchPerformed() {
+        int[][] expected = new int[][]{{0, 1, 2, 3}};
+        String query = "unknown";
+        verifyActionResult(expected, "filterQuery", query);
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java]: Not sure if it's better to inline query, e.g. `verifyActionResult(expected, "filterQuery", "unknown");`

File: src/main/java/teammates/ui/pagedata/AdminEmailLogPageData.java (hunk 1 above)
Comment (xpdavid) [inline src/main/java/teammates/ui/pagedata/AdminEmailLogPageData.java]: The time period is actually controlled by the query in GAE. We don't need to filter it again in this method. (Or we don't want to fetch all logs and filter their time as we can specify time period when fetch logs in GAE)

File: src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java (hunk 4 above)
Comment (xpdavid) [inline src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java]: Maybe for consistency? As there are long queries below. :P

File: src/main/java/teammates/ui/pagedata/AdminEmailLogPageData.java (hunk 1 above)
Comment (whipermr5) [inline src/main/java/teammates/ui/pagedata/AdminEmailLogPageData.java]: I see; good catch!

File: src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java (hunk 4 above)
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java]: I noticed haha; they make sense below as each verify block has multiple lines and the blocks are separated by blank lines, but over here each verify is only one line (without the query string assignment). Either inline them over here (I prefer this as it's more compact) or add a blank line after each `verifyActionResult` to make the code easier to read?

**Comment by CanIHasReview** [comment]
# v2

@xpdavid submitted v2 for review.

* [v2 1/6] [AdminEmailLogPageData: fix bugs: isShowTestData is ignored when query is invalid ](https://github.com/TEAMMATES/teammates/pull/6876/commits/bbeb1fd2f9ad2c738261038aaa3b552ad81be4f1)
* [v2 2/6] [AdminEmailLog: Remove dead code \(cannot find usage of these codes\) ](https://github.com/TEAMMATES/teammates/pull/6876/commits/86084b22796a653f5cd006df0ae0ff1f360f6d52)
* [v2 3/6] [AdminEmailLog: Fix bugs for continue search with query \`after\` and \`before\` ](https://github.com/TEAMMATES/teammates/pull/6876/commits/cb6ca952ec6bed61194d423ca9a4edbc87e532ca)
* [v2 4/6] [AdminEmailLog: Ignore whitespaces in query ](https://github.com/TEAMMATES/teammates/pull/6876/commits/26ad788d98087596f96197c4ca38e41879a71290)
* [v2 5/6] [AdminActivityLogTest: Rename group name to avoid confusion ](https://github.com/TEAMMATES/teammates/pull/6876/commits/50e0a51b9b545da446c8d4bbc1ab9c2b90029d24)
* [v2 6/6] [Implement AdminEmailLogPageActionTest ](https://github.com/TEAMMATES/teammates/pull/6876/commits/7efcba14f48bff651508c65ac3f8ebbdfc3e4b16)

([:books: Archive](https://github.com/TEAMMATES/teammates/compare/e742d5d5d6d19c73f9da203fa137368a916cfd0e...7efcba14f48bff651508c65ac3f8ebbdfc3e4b16)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/TEAMMATES/teammates/compare/fd915f042249de4b50dd30e947b2e3e075a33caf~1...fd915f042249de4b50dd30e947b2e3e075a33caf))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/TEAMMATES/teammates.git refs/pr/6876/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+
+        verifyActionResult(expected, "filterQuery", "unknown");
+
+        verifyActionResult(expected, "filterQuery", "");
+
+        verifyActionResult(expected, "filterQuery", "info");
+
+        verifyActionResult(expected, "filterQuery", "info:");
+
+        String query = "receiver unknown_connector info:content";
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java]: I think either inline all the way or use the query variable all the way from the start in this method? Seems weird to start using it halfway haha, unless this line onwards marks a different logical block?

File: src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+        Date today = TimeHelper.getDateOffsetToCurrentTime(0);
+        insertLogMessagesAtTime(logMessages.get(LOG_MESSAGE_INDEX_TODAY), today.getTime());
+    }
+
+    @Test(groups = "typicalEmailLogs")
+    public void filterQuery_invalidQuery_defaultSearchPerformed() {
+        int[][] expected = new int[][]{{0, 1, 2, 3}};
+
+        verifyActionResult(expected, "filterQuery", "unknown");
+
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java]: Maybe the blank lines can be removed for inlined queries so it's more compact?

File: src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java (hunk 5 above)
Comment (xpdavid) [inline src/test/java/teammates/test/cases/action/AdminEmailLogPageActionTest.java]: :P then I'll change to the one before and add a blank line.

**Comment by CanIHasReview** [comment]
# v3

@xpdavid submitted v3 for review.

* [v3 1/6] [AdminEmailLogPageData: fix bugs: isShowTestData is ignored when query is invalid ](https://github.com/TEAMMATES/teammates/pull/6876/commits/867f0fc28a87de2b645543c011a16e760d2db69f)
* [v3 2/6] [AdminEmailLog: Remove dead code \(cannot find usage of these codes\) ](https://github.com/TEAMMATES/teammates/pull/6876/commits/7ea3978de11c3a1329ffc9c371d3a22e58bdaf76)
* [v3 3/6] [AdminEmailLog: Fix bugs for continue search with query \`after\` and \`before\` ](https://github.com/TEAMMATES/teammates/pull/6876/commits/24512b8a98b9b28bbbb94fabb21b96bd13e0abe8)
* [v3 4/6] [AdminEmailLog: Ignore whitespaces in query ](https://github.com/TEAMMATES/teammates/pull/6876/commits/51aa95bcbd1783663f5d6ef6ae7aa2a68b9abddc)
* [v3 5/6] [AdminActivityLogTest: Rename group name to avoid confusion ](https://github.com/TEAMMATES/teammates/pull/6876/commits/c646a3128131af8ba993db83b4a970e70c7e10c5)
* [v3 6/6] [Implement AdminEmailLogPageActionTest ](https://github.com/TEAMMATES/teammates/pull/6876/commits/565a14606b20b5eb9d615fdb21d1a65ddc4e78b1)

([:books: Archive](https://github.com/TEAMMATES/teammates/compare/9dd911875875a1130549d86a7067b86b0479a74f...565a14606b20b5eb9d615fdb21d1a65ddc4e78b1)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/TEAMMATES/teammates/compare/f7fcc85897a2c71b7baff3aed8bac70715824026~1...f7fcc85897a2c71b7baff3aed8bac70715824026))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/TEAMMATES/teammates.git refs/pr/6876/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/main/java/teammates/ui/pagedata/AdminEmailLogPageData.java [hunk 7]
```diff
@@ -58,6 +58,32 @@ public String getStatusForAjax() {
...
     }
 
+    /**
+     * Returns now if toDate is not present in the query.
+     */
+    public long getToDate() {
+        if (this.q == null || !this.q.isToDateInQuery) {
+            return TimeHelper.now(0.0).getTimeInMillis();
+        } else {
+            return this.q.toDateValue;
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/pagedata/AdminEmailLogPageData.java]: `no-else-return`: this block can be shortened to:
```java
if (condition) {
    return ...;
}
return ...;
```

File: src/main/java/teammates/ui/pagedata/AdminEmailLogPageData.java [hunk 8]
```diff
@@ -58,6 +58,32 @@ public String getStatusForAjax() {
...
+    }
+
+    /**
+     * Returns 0 if fromDate is not present in the query.
+     */
+    public long getFromDate() {
+        if (this.q == null) {
+            return 0;
+        } else {
+            return this.q.fromDateValue;
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/pagedata/AdminEmailLogPageData.java]: Similarly here.

**Pull request merged**